diffrax/integrate.py

```diff
@@ -142,34 +218,107 @@ def body_fun(state, inplace):
+
+            dense_interp = DenseInterpolation(
                 ts=state.dense_ts,
                 ts_size=state.dense_save_index + 1,
                 interpolation_cls=solver.interpolation_cls,
                 infos=state.dense_infos,
                 direction=1,
             )
-            history_vals = []
-            for delay in delays:
-                delay_val = delay(state.tprev, state.y, args)
-                history_val = history.evaluate(delay_val)
-                history_val.append(history_val)
-            history_vals = tuple(history_vals)
 
-            is_vf_wrapper = lambda x: isinstance(x, VectorFieldWrapper)
+            def history_extrapolation_implicit():
+                def _cond_fun(_val):
+                    _, _, _, _, _, pred, step = _val
+                    return (implicit_step & pred) | (
+                        jnp.invert(implicit_step) & (step == 0)
+                    )
 
-            def _apply_history(x):
-                if is_vf_wrapper(x):
-                    vector_field = jtu.Partial(x.vector_field, history=history_vals)
-                    return VectorFieldWrapper(vector_field)
-                else:
-                    return x
+                def _body_fun(_val):
+                    y_prev, _, dense_info, _, _, _, step = _val
+                    is_vf_wrapper = lambda x: isinstance(x, VectorFieldWrapper)
+
+                    def _apply_history(x):
+                        if is_vf_wrapper(x):
+                            vector_field = _HistoryVectorField(
+                                x.vector_field,
+                                t0,
+                                state.tprev,
+                                state.tnext,
+                                dense_info,
+                                dense_interp,
+                                solver.interpolation_cls,
+                                y0_history,
+                                delays.delays,
+                            )
+                            return VectorFieldWrapper(vector_field)
+                        else:
+                            return x
+
+                    terms_ = jtu.tree_map(_apply_history, terms, is_leaf=is_vf_wrapper)
+                    (y, y_error, dense_info, solver_state, solver_result) = solver.step(
+                        terms_,
+                        state.tprev,
+                        state.tnext,
+                        state.y,
+                        args,
+                        state.solver_state,
+                        state.made_jump,
+                    )
 
-            terms_ = jtu.tree_map(_apply_history, terms, is_leaf=is_vf_wrapper)
-            # TODO: write down implicit problem wrt dense_info, using `terms_`
-            (y, y_error, dense_info, solver_state, solver_result) = terms_  # ...
+                    _pred = (((y - y_prev) / y) > delays.eps).any()
+                    _pred = _pred & (step < 10)
```

Comment:
If we hit the maximum number of steps then we should set `y_error` to all infinities, so that we reject the step and try again with a smaller step size. See for example how it's done with the implicit solvers here:
https://github.com/patrick-kidger/diffrax/blob/b8475527eacdba81328130edd7dadd08a0b34063/diffrax/solver/runge_kutta.py#L745

Reply:
sounds good I added a 
```py 
jtu.tree_map(lambda _y_error: jnp.where(step < 10, _y_error, jnp.inf), y_error)
```
after the step in the implicit part